### PR TITLE
feat: add unit tests and allow configuring kubesec url

### PR DESCRIPTION
**Pull request by prit342**
- Update webhook manifests to support admissionregistration.k8s.io/v1 as admissionregistration.k8s.io/v1beta1 has been deprecated since Kubernetes v1.22
- Add unit tests and update to Go version 1.19
- Ensure that Go packages are pointing to github.com/controlplaneio/kubectl-kubesec 
- Fix test manifests and test with K8s version 1.20 to 1.25.0

### CODE REVIEW

**Comment by spiarh** [comment]
Looks like you went the same path as me in #25 to make it work, I'll close my PR. As mentioned by Jack, can you

**Comment by prit342** [comment]
@spiarh vendor directory has already been removed as the part of this PR

**Comment by spiarh** [comment]
> @spiarh vendor directory has already been removed as the part of this PR

indeed, sorry for the noise :wink:

File: deploy/webhook-registration.yaml.tpl.old-api [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  - name: statefulset.admission.kubesc.io
+    clientConfig:
+      service:
+        name: kubesec-webhook
+        namespace: kubesec
+        path: "/statefulset"
+      caBundle: CA_BUNDLE
+    rules:
+      - operations:
+        - CREATE
```
[the remark this answers is not in the document]
Comment (prit342) [reply deploy/webhook-registration.yaml.tpl.old-api]: Thank you, this has now been fixed

File: deploy/webhook-registration.yaml.tpl [hunk 2]
```diff
@@ -1,5 +1,5 @@
 ---
-apiVersion: admissionregistration.k8s.io/v1beta1
+apiVersion: admissionregistration.k8s.io/v1
 kind: MutatingWebhookConfiguration
```
[the remark this answers is not in the document]
Comment (prit342) [reply deploy/webhook-registration.yaml.tpl]: This is now fixed, moved to ValidatingWebhookConfiguration

File: deploy/webhook-registration.yaml.tpl.old-api [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        namespace: kubesec
+        path: "/daemonset"
+      caBundle: CA_BUNDLE
+    rules:
+      - operations:
+        - CREATE
+        - UPDATE
+        apiGroups:
+        - apps
+        - extensions
```
[the remark this answers is not in the document]
Comment (prit342) [reply deploy/webhook-registration.yaml.tpl.old-api]: This is now fixed, deleted the file and removed the `extensions` api Group

File: pkg/webhook/pod_test.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+				logger:   log.Dummy,
+			}
+
+			decoder := serializer.NewCodecFactory(scheme.Scheme).UniversalDecoder()
+
+			pod := &corev1.Pod{}
+
+			if err := runtime.DecodeInto(decoder, []byte(tt.podSpec), pod); err != nil {
+				t.Errorf("unable to convert %q into Pod object - %v", tt.podSpec, err)
+				return
```
[the remark this answers is not in the document]
Comment (prit342) [reply pkg/webhook/pod_test.go:93]: Fixed, using t.Fatalf()

File: pkg/webhook/pod.go [hunk 5]
```diff
@@ -44,17 +44,23 @@ func (d *podValidator) Validate(_ context.Context, obj metav1.Object) (bool, val
...
+		ScanDefinition(buffer)
+
+	//result, err := kubesec.NewClient().ScanDefinition(buffer)
 	if err != nil {
 		d.logger.Errorf("kubesec.io scan failed %v", err)
 		return false, validating.ValidatorResult{Valid: true}, nil
 	}
-	if result.Error != "" {
-		d.logger.Errorf("kubesec.io scan failed %v", result.Error)
+	if result[0].Error != "" {
```
[the remark this answers is not in the document]
Comment (prit342) [reply pkg/webhook/pod.go:67]: Fixed now

File: pkg/webhook/pod.go [hunk 6]
```diff
@@ -44,17 +44,23 @@ func (d *podValidator) Validate(_ context.Context, obj metav1.Object) (bool, val
...
+		return false, validating.ValidatorResult{Valid: true}, nil
+	}
 
 	d.logger.Infof("Scanning pod %s", kObj.Name)
 
-	result, err := kubesec.NewClient().ScanDefinition(buffer)
+	result, err := kubesecv2.NewClient(kubesecScanURL, timeOut).
+		ScanDefinition(buffer)
+
+	//result, err := kubesec.NewClient().ScanDefinition(buffer)
```
[the remark this answers is not in the document]
Comment (prit342) [reply pkg/webhook/pod.go]: Fixed

File: pkg/webhook/deployment.go [hunk 7]
```diff
@@ -24,7 +24,7 @@ type deploymentValidator struct {
 }
 
 func (d *deploymentValidator) Validate(_ context.Context, obj metav1.Object) (bool, validating.ValidatorResult, error) {
```
[the remark this answers is not in the document]
Comment (prit342) [reply pkg/webhook/deployment.go:27]: Yeah should be in the scope of a future PR

File: pkg/webhook/daemonset_test.go [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+          mountPath: /var/log
+      terminationGracePeriodSeconds: 30
+      volumes:
+      - name: varlog
+        hostPath:
+          path: /var/log
+`,
+		},
+		{
+			name:     "Insecure DaemonSet Spec.",
```
[the remark this answers is not in the document]
Comment (prit342) [reply pkg/webhook/daemonset_test.go]: Fixed

File: Dockerfile [hunk 9]
```diff
@@ -1,31 +1,47 @@
...
-RUN chown -R app:app ./
-
-USER app
-
-CMD ["./kubesec"]
+FROM golang:1.19-alpine AS builder
+COPY . /build
+WORKDIR /build
+RUN ls -ltra /build && \
+    apk add --no-cache build-base && \
```
[the remark this answers is not in the document]
Comment (prit342) [reply Dockerfile]: If I don't have build-base I am will get the following error during image build:
`# runtime/cgo
cgo: C compiler "gcc" not found: exec: "gcc": executable file not found in $PATH
FAIL	github.com/controlplaneio/kubesec-webhook/pkg/webhook [build failed]
The command '/bin/sh -c go test -race -v ./... &&     GO111MODULE=on CGO_ENABLED=0 GOOS=linux go build -tags netgo -a -v -o /build/kubesec-webhook /build/cmd/kubesec' returned a non-zero code: 2
`
`ls` has been removed

File: Dockerfile [hunk 10]
```diff
@@ -1,31 +1,47 @@
...
+    -h ${HOMEDIR} \
+    -s "${SHELL}" \
+    -u "${UID}" \
+    "${USER}"
+
+COPY --from=builder /build/kubesec-webhook /app/kubesec
+USER "${USER}"
+WORKDIR "${HOMEDIR}"
+
+ENTRYPOINT [ "/app/kubesec" ]
```
[the remark this answers is not in the document]
Comment (prit342) [reply Dockerfile]: Added new line at the end of file

File: pkg/webhook/daemonset_test.go [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+			wantErr:  false,
+			result:   false, //
```
[the remark this answers is not in the document]
Comment (prit342) [reply pkg/webhook/daemonset_test.go]: Fixed

File: pkg/webhook/daemonset.go [hunk 12]
```diff
@@ -45,17 +45,22 @@ func (d *daemonSetsValidator) Validate(_ context.Context, obj metav1.Object) (bo
...
+		return false, validating.ValidatorResult{Valid: true}, nil
+	}
 
 	d.logger.Infof("Scanning daemonset %s", kObj.Name)
 
-	result, err := kubesec.NewClient().ScanDefinition(buffer)
+	result, err := kubesecv2.NewClient(kubesecScanURL, timeOut).
+		ScanDefinition(buffer)
+
 	if err != nil {
```
[the remark this answers is not in the document]
Comment (prit342) [reply pkg/webhook/daemonset.go:59]: Fixed

**Comment by prit342** [comment]
@06kellyjac  @spiarh  I have also squashed my changes

**Review by spiarh — APPROVED** [review]
LGTM, thanks @prit342

**Review by 06kellyjac — APPROVED** [review]

**Pull request merged**